cdi/src/main/java/org/hibernate/validator/cdi/internal/ValidateableBeanFilter.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.cdi.internal;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import javax.enterprise.inject.spi.WithAnnotations;
+import javax.validation.Constraint;
+import javax.validation.Valid;
+import javax.validation.executable.ValidateOnExecution;
+
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.classhierarchy.ClassHierarchyHelper;
+
+/**
+ * A filter that checks if the passed class has any of these annotations:
+ * <ul>
+ * <li>{@link Valid}</li>
+ * <li>{@link Constraint}</li>
+ * <li>{@link ValidateOnExecution}</li>
+ * </ul>
+ * <p>
+ * on a type level, or on any member, or any parameter of any member. Super classes
+ * and all implemetned interfaces are considered while looking for these annotations.
+ * <p>
+ * The annotation may be applied as a meta-annotation on any annotation considered.
+ * <p>
+ * This filter is needed as {@link WithAnnotations} does not look into class hierarchy
+ * to look for annotations.
+ *
+ * @author Marko Bekhta
+ */
+public class ValidateableBeanFilter implements Predicate<Class<?>> {
+
+	private static final Set<Class<? extends Annotation>> ALLOWED_ANNOTATIONS = Collections.unmodifiableSet(
+			CollectionHelper.asSet( Valid.class, Constraint.class, ValidateOnExecution.class )
+	);
+
+	@Override
+	public boolean test(Class<?> type) {
+		for ( Class<?> clazz : ClassHierarchyHelper.getHierarchy( type ) ) {
+			if ( isAnnotationPresentAnywhere( clazz ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private boolean isAnnotationPresentAnywhere(Class<?> clazz) {
+		Set<Annotation> processedAnnotations = new HashSet<>();
+		// 1. Is present on a type level:
+		if ( isAnnotationPresentIn( clazz.getDeclaredAnnotations(), processedAnnotations ) ) {
+			return true;
+		}
+
+		// 2. Or on a field level
+		for ( Field field : clazz.getDeclaredFields() ) {
+			if ( isAnnotationPresentIn( field.getDeclaredAnnotations(), processedAnnotations ) ) {
```

Comment:
@marko-bekhta Ah ah, when I took my shower this morning, I thought about the type argument constraints/`@Valid`. We should take them into account for fields, parameters and return values.

Always take a shower before merging a PR :).

Other than that, it looks good.

Reply:
:) that's a good point. Wonder if `@WithAnnotations` would pick up such case.

Reply:
Yeah, not so sure, but better get it entirely right this time.